vmm/src/device_manager.rs

```diff
@@ -2016,7 +2086,22 @@ impl DeviceManager {
                 .map_err(DeviceManagerError::CreateVirtioMem)?,
             ));
 
-            devices.push((Arc::clone(&virtio_mem_device) as VirtioDeviceArc, false, id));
+            devices.push((
+                Arc::clone(&virtio_mem_device) as VirtioDeviceArc,
+                false,
+                id.clone(),
+            ));
+
+            // Fill the device tree with a new node. In case of restore, we
+            // know there is nothing to do, so we can simply override the
+            // existing entry.
+            self.device_tree.insert(
+                id.clone(),
+                DeviceNode::new(
+                    id,
+                    Some(Arc::clone(&virtio_mem_device) as Arc<Mutex<dyn Migratable>>),
```

Comment:
It could be worth adding a macro for that, that would take and id and a device as its parameters. It would not bring anything but readability to the code.

Reply:
I might be missing how to do it correctly because I don't see how this would help the readability. From what I understand, if I was creating a macro `new_device_node!()`, each node creation would look very much the same:
```rust
new_device_node!(id.clone(), Arc::clone(&virtio_mem_device) as Arc<Mutex<dyn Migratable>>)
```
Again, I might be missing your point cause I'm not great with macros :)

Reply:
I was thinking about something that would look like:

```Patch
diff --git a/vmm/src/device_manager.rs b/vmm/src/device_manager.rs
index bb30c885..1b946335 100644
--- a/vmm/src/device_manager.rs
+++ b/vmm/src/device_manager.rs
@@ -581,6 +581,15 @@ impl DeviceNode {
     }
 }
 
+macro_rules! device_node {
+    ($id:ident, $device:ident) => {
+        DeviceNode::new(
+            $id.clone(),
+            Some(Arc::clone(&$device) as Arc<Mutex<dyn Migratable>>),
+        )
+    };
+}
+
 #[derive(Clone, Serialize, Deserialize)]
 struct DeviceTree(HashMap<String, DeviceNode>);
 
@@ -953,13 +962,8 @@ impl DeviceManager {
                 // Fill the device tree with a new node. In case of restore, we
                 // know there is nothing to do, so we can simply override the
                 // existing entry.
-                self.device_tree.insert(
-                    iommu_id.clone(),
-                    DeviceNode::new(
-                        iommu_id.clone(),
-                        Some(Arc::clone(&device) as Arc<Mutex<dyn Migratable>>),
-                    ),
-                );
+                self.device_tree
+                    .insert(iommu_id.clone(), device_node!(iommu_id, device));
 
                 (Some(device), Some(mapping))
             } else {
@@ -1578,13 +1582,8 @@ impl DeviceManager {
             // Fill the device tree with a new node. In case of restore, we
             // know there is nothing to do, so we can simply override the
             // existing entry.
-            self.device_tree.insert(
-                id.clone(),
-                DeviceNode::new(
-                    id.clone(),
-                    Some(Arc::clone(&vhost_user_net_device) as Arc<Mutex<dyn Migratable>>),
-                ),
-            );
+            self.device_tree
+                .insert(id.clone(), device_node!(id, vhost_user_net_device));
```

Reply:
Ah yes that looks better indeed :)
Also, I can create another macro that would take only an `id` for the cases where we provide the `None`.